In [160]:
import pandas as pd
import numpy as np

In [142]:
path_file = "./negative_noCDS_relationship/negative_noCDS_relationship_named.csv"
df = pd.read_csv(path_file, sep=",", header=None)
print(df.shape)
print(df.dtypes)
df.head()

(487, 14)
0     object
1     object
2     object
3     object
4     object
5     object
6     object
7     object
8     object
9     object
10    object
11    object
12    object
13    object
dtype: object


0    1    2    3    4    5    6    7   \
0  Seq_0_LinJ.01|rejected_noCDS_c01.10  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  Seq_1_LinJ.01|rejected_noCDS_c01.20  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
2  Seq_2_LinJ.01|rejected_noCDS_c01.30  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
3  Seq_3_LinJ.01|rejected_noCDS_c01.40  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
4  Seq_4_LinJ.01|rejected_noCDS_c01.50  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

    8    9    10   11   12   13  
0  NaN  NaN  NaN  NaN  NaN  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN

In [146]:
# Let's make a list of all the values in the pd dataframe without NaN values
values = df.values.flatten().tolist()
values = [value for value in values if pd.notnull(value)]
values = sorted(values, key=lambda x: int(x.split("_")[1]))
print(len(values))
print(values)

721
['Seq_0_LinJ.01|rejected_noCDS_c01.10', 'Seq_1_LinJ.01|rejected_noCDS_c01.20', 'Seq_2_LinJ.01|rejected_noCDS_c01.30', 'Seq_3_LinJ.01|rejected_noCDS_c01.40', 'Seq_4_LinJ.01|rejected_noCDS_c01.50', 'Seq_5_LinJ.01|rejected_noCDS_c01.60A', 'Seq_6_LinJ.02|rejected_noCDS_c02.70', 'Seq_7_LinJ.02|rejected_noCDS_c02.80', 'Seq_8_LinJ.02|rejected_noCDS_c02.90B', 'Seq_9_LinJ.02|rejected_noCDS_c02.100', 'Seq_10_LinJ.02|rejected_noCDS_c02.110', 'Seq_11_LinJ.02|rejected_noCDS_c02.120', 'Seq_12_LinJ.02|rejected_noCDS_c02.130B', 'Seq_13_LinJ.02|rejected_noCDS_c02.140', 'Seq_14_LinJ.02|rejected_noCDS_c02.150', 'Seq_15_LinJ.02|rejected_noCDS_c02.160B', 'Seq_16_LinJ.02|rejected_noCDS_c02.170', 'Seq_17_LinJ.03|rejected_noCDS_c03.180', 'Seq_18_LinJ.03|rejected_noCDS_c03.190', 'Seq_19_LinJ.04|rejected_noCDS_c04.200C', 'Seq_20_LinJ.04|rejected_noCDS_c04.210D', 'Seq_21_LinJ.04|rejected_noCDS_c04.220D', 'Seq_22_LinJ.04|rejected_noCDS_c04.230', 'Seq_23_LinJ.04|rejected_noCDS_c04.240', 'Seq_24_LinJ.04|rejecte

Why the hell the length of the elements is 721 when the real negative database is 723?!

In [147]:
# Let's get a list with the numbers in "Seq_{x}_LinJ". Only the "x"
numbers = [int(value.split("_")[1]) for value in values]
print(numbers)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221,

In [148]:
# Check which IDs doesn't appear
for i in range(722+1):
    if i not in numbers:
        print(i)

335
533


In [149]:
# Load the original database corrected
path_db = "./negative_noCDS_relationship/negative_database_nomatch_corrected.csv"
og_db = pd.read_csv(path_db, sep=",", header=0)
print(og_db.shape)
print(og_db.dtypes)
og_db.head()

(723, 5)
sseqid     object
sstart      int64
send        int64
sstrand    object
sseq       object
dtype: object


sseqid  sstart    send sstrand  \
0  LinJ.01   36103   36242    plus   
1  LinJ.01  113760  114388    plus   
2  LinJ.01  146412  146530    plus   
3  LinJ.01  261866  262439    plus   
4  LinJ.01  271363  271650    plus   

                                                sseq  
0  AGACAGACCGACACACGCAGCCGTGTGATGCCGCCGCCGAGGGCAG...  
1  CAGCGCCATGCACGACATGGCCGCTGACGTCCGTAGCCCTAACTCG...  
2  GCGAATTGTGTTCTGCGCATGCCTCTTCTCTGCCGTGCAGCATGCG...  
3  CGGACTTGGCAAGTGGCCGCCATCGATGAAAACGCACCATGCCTTT...  
4  CGAACGCCGCCCTCAATCGCGCGCTGAACTTCACGCGGCGGTCGAC...

In [161]:
# Let's add a column with np.nan values
og_db["name"] = np.nan
og_db.head()

sseqid  sstart    send sstrand  \
0  LinJ.01   36103   36242    plus   
1  LinJ.01  113760  114388    plus   
2  LinJ.01  146412  146530    plus   
3  LinJ.01  261866  262439    plus   
4  LinJ.01  271363  271650    plus   

                                                sseq  name  
0  AGACAGACCGACACACGCAGCCGTGTGATGCCGCCGCCGAGGGCAG...   NaN  
1  CAGCGCCATGCACGACATGGCCGCTGACGTCCGTAGCCCTAACTCG...   NaN  
2  GCGAATTGTGTTCTGCGCATGCCTCTTCTCTGCCGTGCAGCATGCG...   NaN  
3  CGGACTTGGCAAGTGGCCGCCATCGATGAAAACGCACCATGCCTTT...   NaN  
4  CGAACGCCGCCCTCAATCGCGCGCTGAACTTCACGCGGCGGTCGAC...   NaN

We can see how there are 723 elements!!

In [163]:
og_db['name'] = np.nan  # This line initializes the column with NaN, which is of dtype float64

# Change the dtype of 'name' column to object (which is used for string data)
og_db['name'] = og_db['name'].astype('object')

# Now proceed with your loop
for _, value in enumerate(values):
    new_name = value.split("|")[1]
    index = int(value.split("_")[1])
    og_db.loc[index, 'name'] = new_name

og_db.head()

sseqid  sstart    send sstrand  \
0  LinJ.01   36103   36242    plus   
1  LinJ.01  113760  114388    plus   
2  LinJ.01  146412  146530    plus   
3  LinJ.01  261866  262439    plus   
4  LinJ.01  271363  271650    plus   

                                                sseq                   name  
0  AGACAGACCGACACACGCAGCCGTGTGATGCCGCCGCCGAGGGCAG...  rejected_noCDS_c01.10  
1  CAGCGCCATGCACGACATGGCCGCTGACGTCCGTAGCCCTAACTCG...  rejected_noCDS_c01.20  
2  GCGAATTGTGTTCTGCGCATGCCTCTTCTCTGCCGTGCAGCATGCG...  rejected_noCDS_c01.30  
3  CGGACTTGGCAAGTGGCCGCCATCGATGAAAACGCACCATGCCTTT...  rejected_noCDS_c01.40  
4  CGAACGCCGCCCTCAATCGCGCGCTGAACTTCACGCGGCGGTCGAC...  rejected_noCDS_c01.50

In [173]:
# We now there are two missing values: index 335 y 533
og_db.loc[
    (og_db["name"].isnull()),
    "name" # Column to be updated
] = "rejected_noCDS_BLAST_ERROR"

In [174]:
og_db.loc[
    og_db["name"] == "rejected_noCDS_BLAST_ERROR",
    :
]

sseqid  sstart  send sstrand  \
335  LinJ.27       3   114    plus   
533  LinJ.34     346   490    plus   

                                                  sseq  \
335  CCCTACACCCTAACCCTAACCCTAACCCAGTACACCCAGTACACCA...   
533  ACACCCGTACACCCGTACACCCGTACACCCGTCACCCTACACCCGC...   

                           name  
335  rejected_noCDS_BLAST_ERROR  
533  rejected_noCDS_BLAST_ERROR

In [175]:
# Save database
og_db.to_csv("./negative_noCDS_relationship/relationship_database/negative_database_nomatch_corrected_named.csv", sep=",", index=False, header=True)